xapian-applications/omega/omega.cc

```diff
@@ -174,7 +174,14 @@ try {
 
     hits_per_page = 0;
     auto val = cgi_params.find("HITSPERPAGE");
-    if (val != cgi_params.end()) hits_per_page = atol(val->second.c_str());
+    if (val != cgi_params.end()) {
+	if (val->second[0] < '0' || val->second[0] > '9' ||
+	    val->second[0] == '-') {
+	    cerr << "Value entered is not in range\n"
+		<< "Range is 10-1000";
```

Comment:
This isn't a good way to report a bad parameter value from a CGI program.

In a typical setup with apache, `stderr` from the CGI gets logged to `/var/log/apache2/error.log` which rarely gets looked at in my experience, and it's only visible to the sysadmin, not the person making the bad request.  The CGI specification doesn't seem to mandate a way to handle `stderr`, so this may vary between servers.

The check for `-` is redundant, since you already to ensure it's between `0` and `9` (and the ASCII digits are contiguous).

This also only complains about a subset of bad values - for example, `0xdeadbeaf` is quietly treated as `0`.

Reply:
> This also only complains about a subset of bad values - for example, `0xdeadbeaf` is quietly treated as `0`.
>
@ojwb As you mentioned on IRC that CGI parameters allow text after the number, shouldn't ' `0xdeadbeaf` be a valid value?.

Reply:
> As you mentioned on IRC that CGI parameters allow text after the number

It's documented as allowed for **some** CGI parameters, e.g. from docs/cgiparams.rst:

{{{
If a parameter named '<' or '>' exists, Omega will go to the previous or next results page (based on the value of TOPDOC), respectively. If not, and a parameter named '[' or '#' exists, it will jump to the page number given by that parameter (trailing junk after the number is ignored). (See the section below on modification of CGI parameters to see how this works.)
}}}

So CGI parameters `[` and `#` ought to allow that, as it's clearly documented as supported (or else we would need to decide that wasn't useful any more and explicitly deprecate it in the next release series, then actually drop support in the one after).

There are some others where the behaviour is documented as lax, e.g.:

```
DOCIDORDER
    set the ordering used when a comparison ends up being by docid (i.e. two documents with equal relevance and/or values). By default (if DOCIDORDER isn't set or is empty) this puts them in ASCENDING order (the lowest document id ranks highest). If DOCIDORDER is specified and non-empty it can begin with "D" for DESCENDING order, "A" for ASCENDING order or any other character ("X" by convention) for DONT_CARE (the Xapian database backend will use whichever order is most efficient). Any characters after the first are ignored.
```

Probably the best way to report an error nicely from Omega is to throw a `std::string` (or `const char*` if it's a fixed message), which is caught by the top level `try`/`catch` in omega.cc, and reported as `Exception: ` followed by the message (escaped for use in HTML to avoid cross-site scripting attacks via error messages).

(More generally, when adding error checks to an existing codebase, look around to see if there are any existing error checks of a similar sort, and what they do to report their errors).

You can throw `std::exception` and it'll be caught and reported, but that's more intended for an implicitly thrown `std::exception` (e.g. misuse of the STL), as something like:

```
Exception: COLLAPSE parameter must be >= 0
```

reads better to someone who may well not be familiar with C++ than something like:

```
Exception: std::exception COLLAPSE parameter must be >= 0
```